src/engine/Service.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2022, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Hannah Bast (bast@cs.uni-freiburg.de)
+
+// NOTE: The `util/HttpServer/beast.h` *must* come before one or several of the
+// `<boost/...>` includes because the former defines some variables which, if
+// undefined, are defined the latter. If the `util/HttpServer/beast.h` include
+// comes later (or in another file), this leads to a "redefinition" warning and,
+// much worse, a nasty segmentation fault in the beast code below.
+//
+// TODO: Solve this in a more robust manner. The order of includes should not
+// matter and it should certainly not cause segmentation faults.
+#include "engine/Service.h"
+
+#include <absl/strings/str_cat.h>
+#include <absl/strings/str_split.h>
+
+#include "engine/CallFixedSize.h"
+#include "engine/Values.h"
+#include "engine/VariableToColumnMap.h"
+#include "parser/TokenizerCtre.h"
+#include "parser/TurtleParser.h"
+#include "util/Exception.h"
+#include "util/HashSet.h"
+#include "util/http/HttpClient.h"
+#include "util/http/HttpUtils.h"
+
+// ____________________________________________________________________________
+Service::Service(QueryExecutionContext* qec,
+                 parsedQuery::Service parsedServiceClause)
+    : Operation(qec) {
+  parsedServiceClause_ = parsedServiceClause;
+  using Url = ad_utility::httpUtils::Url;
+
+  // TODO: The following code sends a SPARQL query to the remote endpoint and
+  // turns the TSV result into a `parsedQuery::Values` object. Is this
+  // constructor the right place for this code or should this be in
+  // `computeResult` or somewhere else?
+
+  // Get the URL of the SPARQL endpoint (we can only do fixed IRIs at the
+  // moment).
+  std::string_view serviceIriString = parsedServiceClause_.serviceVarOrIri_;
+  if (serviceIriString.starts_with("?")) {
+    throw std::runtime_error(
+        "SERVICE with variable for endpoint not yet supported by QLever");
+  }
+  AD_CHECK(serviceIriString.starts_with("<") &&
+           serviceIriString.ends_with(">"));
+  serviceIriString.remove_prefix(1);
+  serviceIriString.remove_suffix(1);
+  Url serviceUrl(serviceIriString);
+
+  // Construct the query to be sent to the SPARQL endpoint.
+  std::string serviceQuery =
+      absl::StrCat(parsedServiceClause_.servicePrologue_, "\nSELECT * WHERE ",
+                   parsedServiceClause_.serviceQueryBody_);
+  LOG(INFO) << "Sending SERVICE query to remote endpoint "
+            << "(protocol: " << serviceUrl.protocolAsString()
+            << ", host: " << serviceUrl.host()
+            << ", port: " << serviceUrl.port()
+            << ", target: " << serviceUrl.target() << ")" << std::endl
+            << serviceQuery << std::endl;
+
+  // Send the query and get the result.
```

Comment:
Add a TODO here stating that the TSV result has the disadvantage of not being deterministic for literals/escaped values etc, and that we should probably use some other more complicated format for arbitrary endpoints, and some ID-based formats for other QLever endpoints.